packages/lit-dev-content/src/github/github-signin.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import type {GitHubSigninReceiverMessage} from './github-types.js';
+
+/**
+ * Options for signInToGithub.
+ */
+export interface GitHubSigninOptions {
+  /**
+   * GitHub OAuth App client ID. Generated when a GitHub OAuth App is first
+   * created.
+   */
+  clientId: string;
+
+  /**
+   * URL where users will be redirected to authorize with GitHub.
+   */
+  authorizeUrl: string;
+}
+
+/**
+ * Pop open a window to authenticate the current user with GitHub following the
+ * procedure documented at
+ * https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#web-application-flow.
+ */
+export const signInToGithub = async (
+  options: GitHubSigninOptions
+): Promise<GitHubSigninReceiverMessage> => {
+  const url = new URL(options.authorizeUrl);
+  url.searchParams.set('client_id', options.clientId);
+  url.searchParams.set('scope', 'gist');
+  url.searchParams.set(
+    'redirect_uri',
+    new URL('/playground/signin/', document.location.href).href
+  );
+  // Width and height found by empirically.
+  const width = 600;
+  const height = 650;
+  // Position the popup in the middle of this tab.
+  const top = window.screenY + window.outerHeight / 2 - height / 2;
+  const left = window.screenX + window.outerWidth / 2 - width / 2;
+  const popup = window.open(
+    url,
+    '_blank',
+    `width=${width},height=${height},top=${top},left=${left}`
+  );
+  if (popup === null) {
+    throw new Error('Could not open window. Are popups disabled?');
+  }
+  const result = await new Promise<GitHubSigninReceiverMessage>((resolve) => {
+    const listener = (event: MessageEvent<GitHubSigninReceiverMessage>) => {
+      if (event.source === popup) {
+        resolve(event.data);
+        window.removeEventListener('message', listener);
+      }
+    };
+    window.addEventListener('message', listener);
```

Comment:
Nit: `options.once`? It also probably doesn't matter - but I don't think this will clean up if the user closes the popup without one of the buttons.

Reply:
We can't actually use `once`, because that would remove the event listener as soon as *any* `message` event was fired on window. We can't assume that our popup is the only thing sending post messages, so that's why we have to check `event.source === popup`. Added a comment about this.

Reply:
Ah of course! Makes sense!